Allow for Negative Indexes #34 (Issue Resolved)

TAJ-18 opened the pull request:
This PR introduces support for negative indices in the terminal library, allowing users to reference positions relative to the end of the screen (e.g., -1 for the last row or column). This follows Lua’s convention for negative indices, enhancing usability for cursor positioning, scroll regions, and related functions.

Changes
1) Added Helper Function:
-resolve_index: Converts negative indices to positive values based on terminal dimensions (height or width).

2) Added Terminal Dimension Functions:
-get_height: Retrieves the terminal screen height.
-get_width: Retrieves the terminal screen width.

3) Updated Cursor Positioning:
-Modified cursor_set and cursor_sets to support negative row and column indices.

4) Updated Scroll Region:
-Modified scroll_region to handle negative indices for start_row and end_row.

Testing
Verified correct handling of negative indices for cursor positioning and scroll regions.
Tested edge cases (-1, -2, etc.) to ensure consistent behavior.

Related Issue
Closes #34 (Allow for Negative Indexes)

Tieske commented on src/terminal/init.lua:
```diff
@@ -18,6 +18,26 @@ local M = {
+  return index
+end
+
+--- Returns the height of the terminal screen.
+-- @treturn number the height in rows
+function M.get_height()
+  return sys.get_terminal_size().height
+end
+
+--- Returns the width of the terminal screen.
+-- @treturn number the width in columns
+function M.get_width()
+  return sys.get_terminal_size().width
+end
+
```
I do not like these functions. Making a function call to make a function call is inefficient.

Also, I think in most scenario's you'd need both height & width. If someone would use these instead, the inefficiency would be even bigger.

Tieske commented on src/terminal/init.lua:
```diff
@@ -18,6 +18,26 @@ local M = {
   _DESCRIPTION = "Cross platform terminal library for Lua (Windows/Unix/Mac)",
 }
 
+-- Helper function to resolve negative indices
+local function resolve_index(index, max_value)
+  if index < 0 then
+    return max_value + index + 1
+  end
+  return index
+end
```
a `utils` module was added, would that module be a better location for this function?

Tieske commented on src/terminal/init.lua:
```diff
@@ -331,6 +355,9 @@ end
 -- @treturn string ansi sequence to write to the terminal
 -- @within cursor_position
 function M.cursor_sets(row, column)
+  -- Resolve negative indices
+  row = resolve_index(row, M.get_height())
+  column = resolve_index(column, M.get_width())
```
so this is what I meant with the inefficiencies.

2 functions calls, that will both call `sys.get_terminal_size`, each only using a single property. Instead call that function directly here, and we will save 50% of cpu cycles probably.

Tieske commented on src/terminal/init.lua:
```diff
@@ -340,17 +353,21 @@ end
 -- @within cursor_position
 function M.cursor_set(row, column)
-  output.write(M.cursor_sets(row, column))
+  -- Retrieve terminal dimensions once
+  local size = sys.get_terminal_size()
+  row = utils.resolve_index(row, size.height)
+  column = utils.resolve_index(column, size.width)
+
+  -- Write the cursor position to the terminal
+  output.write("\27[" .. tostring(row) .. ";" .. tostring(column) .. "H")
```
why this change? if the function `cursor_sets` already does the negative-index-update, then we don't need to change this function at all right?

Tieske commented on src/terminal/init.lua:
```diff
@@ -18,14 +18,38 @@ local M = {
+-- @tparam number column
+-- @return true
+-- @within cursor_position
+function M.cursor_set(row, column)
```
this function is overwritten by the implementation below. So this can go, correct? and then the `get_dimensions` above is also no longer used?

Tieske commented on src/terminal/init.lua:
```diff
@@ -18,14 +18,38 @@ local M = {
   unpack = function(t, i, j) return oldunpack(t, i or 1, j or t.n or #t) end
 end
 
-
```
these whitespace changes are unrelated to this PR. I think it is better to not mix multiple issues in 1 PR. Can they be removed?

Tieske commented on src/terminal/init.lua:
```diff
@@ -331,6 +340,10 @@ end
 -- @treturn string ansi sequence to write to the terminal
 -- @within cursor_position
 function M.cursor_sets(row, column)
+ -- Retrieve terminal dimensions once
+local size = sys.get_terminal_size()
+row = utils.resolve_index(row, size.height)
+column = utils.resolve_index(column, size.width)
```
```suggestion
  -- Retrieve terminal dimensions once
  local size = sys.get_terminal_size()
  row = utils.resolve_index(row, size.height)
  column = utils.resolve_index(column, size.width)
```
indentation is of a little bit here

Tieske commented on src/terminal/scroll.lua:
```diff
@@ -29,11 +33,25 @@ function M.scroll_regions(top, bottom)
-  output.write(M.scroll_regions(top, bottom))
+-- @within scroll_region
+function M.scroll_region(start_row, end_row)
+  -- Input validation
+  assert(type(start_row) == "number", "start_row must be a number")
+  assert(type(end_row) == "number", "end_row must be a number")
```
I get what you're doing here, but since we don't check arguments anywhere, should we here?

Tieske commented on src/terminal/scroll.lua:
```diff
@@ -29,11 +33,25 @@ function M.scroll_regions(top, bottom)
+  -- Ensure start_row is less than or equal to end_row
+  assert(start_row <= end_row, "start_row must be less than or equal to end_row")
+
+  output.write("\27[" .. tostring(start_row) .. ";" .. tostring(end_row) .. "r")
```
All of this logic should go in the `scroll_regions` function above, so from here we can just call that function. Similar as the other comment, with `cursor_set`.

Tieske commented on src/terminal/scroll.lua:
```diff
@@ -109,17 +127,23 @@ function M.scroll_apply()
+-- @tparam[opt] number bottom Bottom row of the scroll region (can be negative).
 -- @treturn string ANSI sequence of the new scroll region.
 function M.scroll_pushs(top, bottom)
+  if top and bottom then
+    top = resolve_index(top, get_height())
+    bottom = resolve_index(bottom, get_height())
+  end
```
same again; this work should already be done in `scroll_regions`, so no need to do it again here.

Tieske reviewed (changes requested):
nice work! 👍 
some comments to think about

TAJ-18 commented:
Hi @Tieske , I’ve implemented the suggested improvements—refactored terminal cursor positioning, optimized scroll module, and removed redundant index resolution. Would love your feedback on the changes. Let me know if any further refinements are needed. Thanks for your guidance!

Tieske commented:
@TAJ-18 this PR has now become a mix of many different updates; whitespace, comments, typos, some nice bugfixes, and the actual feature.

Can we close this and create a PR per feature/fix? For some of the small typos etc, it's probably easier to use the github UI, to interactively change it.

That would also fix the conflicts we currently have.
